Add launcher key in slurm object in buildtest configuration

shahzebsiddiqui opened the pull request:
update default.yml
Add slurm-example.yml as reference configuration for sites that using
Slurm
Add executor key in script schema, this refers to name of executor in
configuration to used when dispatching job. This is a required key
Update an example script (examples/script/slurm.yml) to use executor key
Removed the try/except clause in buildtest configuration check, this was
done because validate method prints useful output for troubleshooting
schema, the error message we print to screen was not showing anything
useful.

shahzebsiddiqui committed:
Add launcher key in slurm object in buildtest configuration
update default.yml
Add slurm-example.yml as reference configuration for sites that using
Slurm
Add executor key in script schema, this refers to name of executor in
configuration to used when dispatching job. This is a required key
Update an example script (examples/script/slurm.yml) to use executor key
Removed the try/except clause in buildtest configuration check, this was
done because validate method prints useful output for troubleshooting
schema, the error message we print to screen was not showing anything
useful.

vsoch commented on buildtest/buildsystem/schemas/script/script-v0.0.1.schema.json:
```diff
@@ -3,7 +3,7 @@
   "title": "BuildTest Test Configuration for Script",
   "type": "object",
-  "required": ["type", "run"],
+  "required": ["type", "executor","run"],
```
This is breaking portability. A test recipe should not have a hard coded executor, it’s determined by the runners config or command line specification at runtime. Also, changes to the schema don’t go here, they are reviewed first in the schemas repository.

shahzebsiddiqui replied:
i would prefer having ``executor`` in recipe otherwise it will be problematic especially when it comes to doing directory build (`buildtest build -c dir``) where more than one yml file is processed that may use two different executor. For example, there is one executor for GPU partition, and another for MPI. I have no way of doing this via cli other than running two separate builds. 

Sure, i will make change to same change to schema repo as well.

vsoch replied:
This would be akin to hardcoding an executor into a Snakefile, meaning that if I move my Snakemake to some other environment I can't just run it as is. The entire purpose of having the user define executors for their host and then specify to use them either with default config or a flag is to make these recipes portable. If executor is added, it should be an *optional* value and then the client here needs to have a flag to ignore it so the recipes are still portable.

shahzebsiddiqui replied:
I understand, in that case i would be fine with making executor as an optional argument and have default executor be one of the local executors. I can make that change

vsoch replied:
The default executor should be whatever the use sets it to be in their local configuration (which I agree would be a local executor when first installed).

shahzebsiddiqui replied:
The change has been made in https://github.com/HPC-buildtest/buildtest-framework/pull/249/commits/a5ab2e7d1f47c99fa936c646c4051f4280964b71 so executor is no longer required. I fixed the Travis issue so it should be all green now.

vsoch commented on buildtest/settings/slurm-example.yml:
```diff
@@ -0,0 +1,14 @@
+  normal:
+    type: slurm
+    description: submit jobs with slurm using srun instead of sbatch
+    launcher: sbatch
```
Launcher needs to be under vars as it is an executor specific variable.

shahzebsiddiqui replied:
``vars`` is an array that can have any key/value pair. The ``launcher`` needs to be of enum restricted to few types.  I may misunderstood, the ``vars`` are for defining shell variables, if its for executor specific vars, how do you intend on using it.

vsoch replied:
Okay that makes sense, but you'll also need to update the executor class [here](https://github.com/HPC-buildtest/buildtest-framework/blob/a5ab2e7d1f47c99fa936c646c4051f4280964b71/buildtest/executors/base.py#L105) that is expecting launcher under vars. I'm not sure if your recipes will validate, but if you report that they do, it definitely is easier to have the slurm type validate for launcher. Perhaps there should be a default for sbatch if it's not defined, and then don't require the user to specify it in the config?

And in this case of defining specific variables, it's not clear to me what purpose "vars" and "options" serve?

vsoch replied:
If "vars" refers to shell variables then they would better be called environment or something like that. Vars is way too general to be clear to the user what it's referring to (I still don't get it).

shahzebsiddiqui replied:
How about let's remove ``vars`` it may be confusing both of us and I dont have a clear idea how it will be used yet

shahzebsiddiqui replied:
the ``options`` is a list of ``SBATCH`` directives that can get injected into the test script. And there can be way to override some of the options in the test configuration. buildtest should add ``#SBATCH`` in front of each item, so that it is used as intended, user shouldn't put any arbitrary value in here.

vsoch replied:
Removing vars sounds good to me! Environment is fairly straight forward, but we have it represented in the [test config](https://github.com/HPC-buildtest/schemas/blob/master/script/script-v0.0.1.schema.json#L27). We'll need to think more about this setup - for example, hard coding environment variables into the testing file isn't the best idea if you want to put them on GitHub. It might also be the case that the runner would want to specify a file name instead, or have a config environment section to override the local one...

Anyway I agree we don't have a good sense of these things, so let's remove vars for now, both for the default config, and the get_variables function.  So you probably want something like:

```python
    def load(self, name):
        """Load the executor preferences from the provided config, which is
           added and indexed with "name." For slurm we look for the following
           in vars:
           :param launcher: defaults to sbatch
           :type launcher: string
        """
        self.launcher = self._settings.get("launcher", "sbatch")
```

shahzebsiddiqui replied:
Let's see if Travis passes all regtest. It worked locally. I made all the changes you requested.

shahzebsiddiqui committed:
need executor in example yml for regtest.
Black all code

shahzebsiddiqui committed:
remove "executor" from required list.
Travis was erroring on  buildtest configuration check, so had to
rewrite the test.

shahzebsiddiqui committed:
remove vars and alphanumeric from json definition
update schema examples and ones used in regression test
Update load method in SlurmExecutor to get launcher variable

vsoch commented on tests/config_schema_examples/slurm-example.yml:
```diff
@@ -1,8 +1,13 @@
     vars:
-      launcher: srun
+      FOO: BAR
+    options:
```
Where are options parsed, and why are there vars here if we removed them?

vsoch replied:
If vars was in the example, why did tests still pass before?

shahzebsiddiqui replied:
it's bit odd since ``additionalProperties: False`` in slurm object https://github.com/shahzebsiddiqui/buildtest-framework/blob/update_schema/buildtest/settings/default.schema.json#L48 i have no idea

shahzebsiddiqui replied:
I didn't parse ``options`` yet, that was something i plan to do in later PR. The #SBATCH directives would be read from the options key and injected into the test, so this would require testing on my end at stampede2.

vsoch replied:
I think it's an issue with the nested ref, this looks like something that might work! https://stackoverflow.com/questions/58170135/how-to-use-additionalproperties-from-a-ref-file/58171051#58171051

shahzebsiddiqui replied:
I looked at the link, it's bit confusing according to the docs https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-6.5.6 

This looks like a silent bug that doesn't show up during testing, and could be something to address in a separate PR. If i understand, we dont want any extra properties passed to any object   whether it be a child object like executor or the parent objects.

vsoch replied:
Basically, the properties serve as booleans to reference if a definition is present or not (and according to the author) this fixes the bug that nesting of schemas doesn't honor the attribute. It's worth a try.

shahzebsiddiqui replied:
so should i remove the following ``additionalProperties`` from ``local`` and ``slurm``. 
https://github.com/shahzebsiddiqui/buildtest-framework/blob/update_schema/buildtest/settings/default.schema.json#L48
https://github.com/shahzebsiddiqui/buildtest-framework/blob/update_schema/buildtest/settings/default.schema.json#L29

I guess the metric for success would be to get the validate method fail when ``vars`` is specified in executor object. Is that right?

vsoch replied:
Yes we want the additionalProperties: false to be honored - in the case of vars it should have triggered an invalid error, but it didn't. It will be important beyond vars, because it could be that a user has a spelling mistake (or something like that) that should be known.

shahzebsiddiqui replied:
Let me do some testing and see what i can fix.

shahzebsiddiqui replied:
so i did a few tests

Test on outer level
---------------------
Expected Result: Fail on ``x: 1``
```
x: 1
editor: vi
executors:
  runlocal:
    type: local
    description: submit jobs locally
```

```
On instance:
    {'editor': 'vi',
     'executors': {'runlocal': {'description': 'submit jobs locally',
                                'type': 'local'},
                   'x': {'type': 'local'}},
     'x': 1}
```
Test inside executor
-----------------------

Expected Result: Fail on ``vars: 1`` not an object 

```
editor: vi
executors:
  runlocal:
    type: local
    description: submit jobs locally
  x:
    type: local
  vars: 1
```

```
Failed validating 'type' in schema[0]:
    {'additionalProperties': False,
     'properties': {'description': {'type': 'string'},
                    'options': {'items': {'type': 'string'},
                                'type': 'array'},
                    'type': {'enum': ['local'], 'type': 'string'}},
     'required': ['type'],
     'type': 'object'}

On instance:
    1
```

Test inside executor object
------------------------------

Expected Result: Fail on ``x: 1``

```
editor: vi
executors:
  runlocal:
    type: local
    description: submit jobs locally
    x: 1
```

What i dont understand is why validate is thinking it needs to match with ``slurm`` object when a new key is being added. Maybe it resorts to the last object since it is using ``anyOf``

```
jsonschema.exceptions.ValidationError: 'local' is not one of ['slurm']

Failed validating 'enum' in schema[1]['properties']['type']:
    {'enum': ['slurm'], 'type': 'string'}

On instance['type']:
    'local'
```

I get same result when i change above schema from ``type: local`` to ``type: slurm``.

Now trying the combined schema from regtest but swap the two executor names and put ``vars: 1`` and it gets same result

```
editor: vi
executors:
  normal:
    type: slurm
    description: submit jobs with slurm using srun instead of sbatch
    launcher: sbatch
    options:
      - "-p normal"
      - "-t 00:10"
      - "-N 1"
      - "-n 1"
  local:
    type: local
    description: submit jobs locally
    vars: 1
```

```
jsonschema.exceptions.ValidationError: 'local' is not one of ['slurm']

Failed validating 'enum' in schema[1]['properties']['type']:
    {'enum': ['slurm'], 'type': 'string'}

On instance['type']:
    'local'
```

Now this last one is a bit interesting, apparently I can define two executors of same name ``local`` and the validate passes

```
editor: vi
executors:
  normal:
    type: slurm
    description: submit jobs with slurm using srun instead of sbatch
    launcher: sbatch
    options:
      - "-p normal"
      - "-t 00:10"
      - "-N 1"
      - "-n 1"
  local:
    type: local
    description: submit jobs locally

  local:
    type: slurm
    launcher: sbatch
    description: submit jobs locally                                      
```

For now i removed the two lines as i referenced in previous message.

shahzebsiddiqui replied:
all my test were conducted with the following commit https://github.com/HPC-buildtest/buildtest-framework/pull/249/commits/3c1b5d8a7386d0aaeedf059e9523f6112bcbc299

shahzebsiddiqui replied:
@vsoch what your thoughts regarding my results? Do you think the issue is resolved?

vsoch replied:
I don't see a case where you add additional variables to an executor and have the expected failure, so I don't think so. If you want me to give it a shot I can make time for it this week.

shahzebsiddiqui replied:
I think that would be good use-case of testing bad configuration examples as part of regression test. Maybe put them in this directory https://github.com/HPC-buildtest/buildtest-framework/tree/devel/tests/config_schema_examples 

That can be handled in a separate PR.

vsoch replied:
Yes, but the issue we were addressing is having undefined fields in the executors, you asked me if that is resolved, and it is not.

vsoch replied:
@shahzebsiddiqui for the last case, when you load the schema it will actually only have *one* of those executors because it's read as a dictionary. There might be a way to prevent duplicate keys but I haven't found it yet - we couldn't check for it in python because it would only load one of the sections.

shahzebsiddiqui replied:
If I may let's document and handle testing of configuration check in separate issue. It's out of scope and not the intent when I was issuing the PR.

shahzebsiddiqui committed:
remove vars from example config for regtest

shahzebsiddiqui committed:
remove additionalProperties from child object 'local' and 'slurm'

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>
